src/plugins/dashboard/server/saved_objects/dashboard_migrations.ts

```diff
@@ -101,17 +101,7 @@ const migrations700: SavedObjectMigrationFn<any, any> = (doc): DashboardDoc700To
 };
 
 export const dashboardSavedObjectTypeMigrations = {
-  /**
-   * We need to have this migration twice, once with a version prior to 7.0.0 once with a version
-   * after it. The reason for that is, that this migration has been introduced once 7.0.0 was already
-   * released. Thus a user who already had 7.0.0 installed already got the 7.0.0 migrations below running,
-   * so we need a version higher than that. But this fix was backported to the 6.7 release, meaning if we
-   * would only have the 7.0.1 migration in here a user on the 6.7 release will migrate their saved objects
-   * to the 7.0.1 state, and thus when updating their Kibana to 7.0, will never run the 7.0.0 migrations introduced
-   * in that version. So we apply this twice, once with 6.7.2 and once with 7.0.1 while the backport to 6.7
-   * only contained the 6.7.2 migration and not the 7.0.1 migration.
-   */
-  '6.7.2': flow(migrateMatchAllQuery),
```

Comment:
I am not sure if we "can" remove (or should) old migrations. The problem is, that depending on from which version you then came, you'll either ran in 6.7.2 over that migration or not. E.g. you were updating from 6.6.0 to 7.6.0 and then 7.10.0, will give you different migrations then upgrading from 6.6.0 to 7.10.0. @rudolf Do you know if this is at all possible? If it's possible I still would like to hear your opinion on if we should do something like that, because this sounds like a horrible source for errors that we'll never be able to find later, since we removed old migrations from the code and won't see it when debugging?

Reply:
@timroes you highlighted a good question in general. It seems to me that wherever we add this migration script, there will still be questions.
Also, analyzing this particular script, I see no problems if it is executed in `7.8.` On the other hand, it seems strange to me to add in `7.8` something that is valid for an upgrade from version `5` to `6`

From my point of view we should keep version as it (**6.7.2**) and just a fix reason of an exception:
![image](https://user-images.githubusercontent.com/20072247/94407454-1e3fc400-017c-11eb-9c55-e49f3dfddc41.png)

Reply:
My understanding is that this migration fixes a bug introduced in v5 (or a bug in the v6 migration). The version the data bug was introduced is irrelevant, the migration version captures when this migration was applied and the data was transformed and fixed. So a search document with a migrationVersion of 7.9.3 means that this data has been migrated and fixed in Kibana 7.9.3. And then we know that any kibana > 7.9.3 can safely consume this document knowing it has the expected shape.

What complicates matters here is that we first released this migration in v7.8.0 as "6.7.2". As Joe pointed out in the original PR this means that this migration would not have been applied to documents which for instance had already been upgraded to v7.4.0 because we only apply newer migrations. So if someone created a v5 document with this bug (or however you reproduce the original issue) and then upgraded through some path to v7.4.0 this transformation will not be applied when they later upgrade to > v7.8.0

So there are two issues which should be addressed, one is correctly handling invalid JSON, the other is making sure that this migration is applied to all documents. To get this migration applied to all versions we need to set it's version to the oldest Kibana this will get released in (e.g. 7.9.3).

Since the "new" 7.9.3 migration is the same as the 6.7.2 migration, anyone running Kibana > 7.9.3 will always have this migration applied (perhaps it's applied twice, but that's fine).

However removing the migration might still be a bad idea because it makes it harder to find and debug because it's no longer in the `master` codebase. So I can see some benefit in keeping the 6.7.2 migration just as a historic reference so that if someone ever wonders why a document had a migration applied with v6.7.2 they'll be able to find that migration in the codebase. We would then have two identical migration scripts with different versions.

Reply:
The other option is to not have duplicate migrations but just have a big comment block in the code above the 7.9.3 migration stating that at some point this was released as a 6.7.2 migration and later changed.

Reply:
I think the "traceability" here would outrule that small performance drawback of running that other (same) migration in 6.7.2 and as far as I understand there'll be no other drawback to it, why I'd suggest we're not removing it, but instead keeping it for the old version.

Reply:
Agreed, I also prefer the better traceability of duplicate migrations.